source/threading_and_code_execution.md

```diff
@@ -0,0 +1,218 @@
+
+Перезагружатель (Reloader)
+--------------------------
+
+Как и исполнитель, перезагружатель также оборачивает код приложения. Если исполнитель уже не активен в текущем треде, перезагружатель вызовет его сам, так что необходимо вызывать только перезагружатель. Это также гарантирует, что все, что делает перезагружатель, включая все его вызовы коллбэка, оказывается обернутым внутри исполнителя.
+
+```ruby
+Rails.application.reloader.wrap do
+  # здесь вызывается код приложения
+end
+```
+
+Перезагружатель подходит только тогда, когда долговременный процесс на уровне фреймворка повторно вызывается в коде приложения, например, для веб-сервера или очереди заданий. Rails автоматически оборачивает веб-запросы и воркеры Active Job, поэтому редко приходится ссылаться на перезагружатель. Всегда учитывайте, подходит ли исполнитель для конкретного случая.
+
+### Коллбэки
+
+Перед входом в обернутый блок перезагружатель проверит, нужно ли перезагружать запущенное приложение - например, потому что исходный файл модели был модифицирован. Если он определяет, что требуется перезагрузка, он будет ждать до тех пор, пока это будет безопасно, а затем сделает это, прежде чем продолжить. Когда приложение настроено на постоянную перезагрузку независимо от того, обнаружены ли какие-либо изменения, перезагрузка будет выполняется в конце блока.
+
+Перезагружатель также предоставляет коллбэки `to_run` и `to_complete`; они вызываются в тех же точках, что и для исполнителя, но только когда текущее выполнение инициировало перезагрузку приложения. Когда нет необходимости в перезагрузке, перезагружатель будет вызывать обернутый блок без каких-либо других коллбэков.
+
+### Выгрузка классов (Class Unload)
+
+Наиболее значительная часть процесса перезагрузки - это выгрузка классов, в которой все автозагруженные классы удаляются, и их можно снова загружать. Это происходит непосредственно до коллбэков `to_run` или `to_complete`, в зависимости от параметра `reload_classes_only_on_change`.
+
+Зачастую дополнительные экшны по перезагрузке должны выполняться как непосредственно до, так и сразу после выгрузки классов, поэтому перезагружатель также предоставляет коллбэки `before_class_unload`
+и `after_class_unload`.
+
+### Конкурентность
+
+Только долговременные процессы "верхнего уровня" должны ссылаться на перезагружатель, потому что если он определяет, что требуется перезагрузка, он будет блокироваться до тех пор, пока все другие треды не завершат любые вызовы исполнителя.
+
+Если это произойдет в "дочернем" треде, с ожидающим родителя внутри исполнителя, это вызовет неизбежный дедлок: перезагрузка должна произойти до того, как дочерний тред будет выполнен, но он не может быть безопасно выполнен, пока родительский тред находится в середине выполнения. Вместо этого дочерний тред должен использовать исполнителя.
```

Comment:
с ожидающим родителя внутри исполнителя ?

> If this were to occur in a "child" thread, with a waiting parent inside the Executor, it would cause an unavoidable deadlock: